crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip/bloom.rs

```diff
@@ -47,76 +49,111 @@ impl ShardedGossipLocal {
     pub(super) async fn generate_ops_blooms_for_time_window(
         &self,
         common_arc_set: &Arc<DhtArcSet>,
-        mut search_time_window: TimeWindow,
-    ) -> KitsuneResult<Vec<TimedBloomFilter>> {
-        let mut results = Vec::new();
-        loop {
-            // Get the op hashes which fit within the common arc set from these local agents.
-            let result = store::all_op_hashes_within_arcset(
-                &self.evt_sender,
-                &self.space,
-                (**common_arc_set).clone(),
-                search_time_window.clone(),
-                Self::UPPER_HASHES_BOUND,
-                true,
-            )
-            .await?;
+        search_time_window: TimeWindow,
+    ) -> KitsuneResult<Batch<TimedBloomFilter>> {
+        use futures::TryStreamExt;
 
-            // If there are none then don't create a bloom.
-            let (ops_within_common_arc, found_time_window) = match result {
-                Some(r) => r,
-                None => {
-                    if !common_arc_set.is_empty() {
-                        // We have local agents within the arc but no hashes.
-                        let bloom = TimedBloomFilter {
-                            bloom: None,
-                            time: search_time_window,
-                        };
-                        results.push(bloom);
-                    }
-                    break;
-                }
-            };
+        // If the common arc set is empty there's no
+        // blooms to generate.
+        if common_arc_set.is_empty() {
+            return Ok(Batch::Complete(Vec::with_capacity(0)));
+        }
 
-            let num_found = ops_within_common_arc.len();
+        let mut total_blooms = 0;
+        let search_end = search_time_window.end;
 
-            // Create the bloom from the op hashes.
-            let mut bloom =
-                bloomfilter::Bloom::new_for_fp_rate(ops_within_common_arc.len(), Self::TGT_FP);
+        let stream = store::hash_chunks_query(
+            self.evt_sender.clone(),
+            self.space.clone(),
+            (**common_arc_set).clone(),
+            search_time_window.clone(),
+            true,
+        );
+        let batch = stream
+            // Take more chunks while there is less then
+            // the upper limit for number of blooms.
+            .try_take_while(|_| {
+                total_blooms += 1;
+                futures::future::ready(Ok(total_blooms <= Self::UPPER_BLOOM_BOUND))
+            })
```

Comment:
Is that different than just this?

```suggestion
            .take(Self::UPPER_BLOOM_BOUND)
```

Reply:
Unfortunately there's no` take` for a [TryStream](https://docs.rs/futures/latest/futures/stream/trait.TryStreamExt.html)